codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AssembleMiddlewareStack.java

```diff
@@ -41,61 +41,91 @@ public byte getOrder() {
+                            }
+                            return false;
+                        })
                         .registerMiddleware(MiddlewareRegistrar.builder()
-                                .resolvedFunction( SymbolUtils.createValueSymbolBuilder(
-                                        "AddRetryMiddlewares", AwsGoDependency.AWS_RETRY)
+                                .resolvedFunction(SymbolUtils.createValueSymbolBuilder(
+                                        "AddComputePayloadSHA256Middleware", AwsGoDependency.AWS_SIGNER_V4)
                                         .build())
-                                .useClientOptions()
                                 .build())
                         .build(),
 
-                // Add unsigned payload middleware to operation stack
+                // Add content-sha256 payload header middleware to operation stack
                 RuntimeClientPlugin.builder()
                         .operationPredicate((model, service, operation) -> {
-                            if (operation.hasTrait(UnsignedPayloadTrait.class)) {
+                            if (service.hasTrait(SigV4Trait.class) && operation.hasTrait(UnsignedPayloadTrait.class)) {
                                 return true;
                             }
                             return false;
                         })
                         .registerMiddleware(MiddlewareRegistrar.builder()
                                 .resolvedFunction(SymbolUtils.createValueSymbolBuilder(
-                                        "AddUnsignedPayloadMiddleware", AwsGoDependency.AWS_SIGNER_V4)
+                                        "AddContentSHA256HeaderMiddleware", AwsGoDependency.AWS_SIGNER_V4)
+                                        .build())
+                                .build())
+                        .build(),
+
+                // Add newAttempt middleware to operation stack
+                RuntimeClientPlugin.builder()
+                        .registerMiddleware(MiddlewareRegistrar.builder()
+                                .resolvedFunction(SymbolUtils.createValueSymbolBuilder(
+                                        "AddRetryMiddlewares", AwsGoDependency.AWS_RETRY)
                                         .build())
+                                .useClientOptions()
                                 .build())
                         .build(),
 
                 // Add HTTPSigner middleware to operation stack
                 RuntimeClientPlugin.builder()
                         .operationPredicate((model, service, operation) -> {
-                            if (service.hasTrait(SigV4Trait.class) && (!operation.hasTrait(UnsignedPayloadTrait.class))
+                            if (service.hasTrait(SigV4Trait.class)
                                     && (operation.hasTrait(SigV4Trait.class) || !operation.hasTrait(AuthTrait.class))
-                            ){
+                            ) {
```

Comment:
Does AuthTrait.class also include UnsignedPayloadTrait.class? Would this predicate work if service is decorated with SigV4 trait and an operation is decorated with `UnsignedPayloadTrait`?

Reply:
No so UnsignedPayloadTrait is a standalone trait that is not coupled to the authentication scheme. `[auth trait](https://awslabs.github.io/smithy/1.0/spec/core/auth-traits.html#auth-trait) is a list of authentication schemes that are intended to apply direclty to the service or subset of auth schemes for an operation.

I think the way the `SigV4Trait` is used on the operation is wrong here, and should be updated to check the values of the `AuthTrait` instead.  It looks like [authDefintionTrait](https://awslabs.github.io/smithy/1.0/spec/core/auth-traits.html#authdefinition-trait) aka Auth schemes are only supposed to decorate services not operations.

Looks like the logic for these middleware needs to be tweeked with the expectation of decorating traits, and auth schemes.

Reply:
switched to use the `ServiceIndex` and `getEffectiveAuthSchemes` utilities to check for auth scheme.